### PR TITLE
Space-traveller-Hub

### PR DESCRIPTION
**Pull request by agneta1991**
## Project Pull Request: Space-Traveller-Hub Review

**Authors:** Agneta Pupienyte, Steven Ochieng

### Description
This pull request is for the review of our SpaceX App project. The project is built using React, Redux, and React Router, and it integrates with SpaceX APIs to fetch and display data about rockets and missions. The work was divided with Agneta working on rocket-related tasks and Steven handling mission-related tasks. The profile section was done collaboratively. We've also set up testing, adhered to best practices, and documented our work.

### Changes Made
- Created React App using Create React App.
- Integrated React Redux, Redux Logger, and React Router.
- Fetched and displayed data from SpaceX API endpoints for rockets and missions.
- Implemented rendering of rockets and missions.
- Developed functionality for reserving and canceling rockets and joining/leaving missions.
- Styled components using separate files for each.
- Set up testing using React Testing Library and Jest.

### Additional Notes
- The project adheres to best practices for HTML/CSS and JavaScript.
- The application design follows the specified requirements.
- All components and functions have been tested.
- We've avoided state mutations by updating the state immutably.

### Checklist
- [x] No linter errors.
- [x] Used correct Gitflow.
- [x] Documented work professionally.
- [x] Followed best practices for HTML, CSS, and JavaScript.
- [x] Fetched and displayed data from SpaceX APIs.
- [x] Implemented reservation and joining functionality using Redux.
- [x] Created conditional rendering for UI components.
- [x] Styled components according to design.
- [x] Implemented My Profile section.
- [x] Set up testing using React Testing Library and Jest.
- [x] Avoided state mutations.

### CODE REVIEW

File: src/redux/rocketSlice.js
Comment (cynthiainga) [inline src/redux/rocketSlice.js:1]: - [ ] As mentioned in the project requirements all Redux state slice files should be created in their own directory which means `rocketSlice` should be in the `rockets` directory 🙃 
- [ ] Please kindly do not disable linters errors, fixing linters helps us to write a clean and strong code base 😁

File: src/redux/missions/missionsSlice.js [hunk 1]
```diff
@@ -0,0 +1,72 @@
+/* eslint-disable max-len */
+// missionsSlice.js
+
+import { createSlice, createAsyncThunk } from '@reduxjs/toolkit';
+import { createSelector } from 'reselect';
+
+const missionsUrl = 'https://api.spacexdata.com/v3/missions';
+
+export const getMissions = createAsyncThunk('missions/getMissions', async () => {
+  // eslint-disable-next-line no-useless-catch
```
Comment (cynthiainga) [inline src/redux/missions/missionsSlice.js]: - [ ] Please kindly do not disable linters, fixing linters helps us to write a strong and clean code base 😁

File: src/components/missions.js [hunk 2]
```diff
@@ -0,0 +1,73 @@
+/* eslint-disable jsx-a11y/control-has-associated-label */
+import React, { useEffect } from 'react';
+import { useDispatch, useSelector } from 'react-redux';
+import {
+  getMissions, selectMissionsStatus, joinMission, leaveMission, selectMappedMissions,
+} from '../redux/missions/missionsSlice';
+// import './missionsTable.css';
```
Comment (cynthiainga) [inline src/components/missions.js]: - [ ] Please kindly do not disable linters, fixing linters helps us to write a strong and clean code base 😁 
- [ ] Kindly delete the commented line of code to keep your code clean 🙃

File: src/styles/Rockets.css
Comment (cynthiainga) [inline src/styles/Rockets.css:1]: - [ ] Please kindly make sure to follow the given design 🙃 
  - Kindly make sure the `reserve` button has the same color as the button in the original design, also in the original design the button doesn’t have a border 😁 
  - Kindly make sure the `reservation status` has the same color as the one in the given design 🙃 

Your design

Screenshot 2023-08-31 at 16 53 50

Given design

Screenshot 2023-08-31 at 16 31 37

File: src/components/missions.css
Comment (cynthiainga) [inline src/components/missions.css:1]: - [ ] Good job so far, but you should make sure your design matches the given design "colors, padding, margin,..."

Your design

Screenshot 2023-08-31 at 16 54 05

Given design

Screenshot 2023-08-31 at 16 31 52

**Review by cynthiainga — CHANGES_REQUESTED** [review]
Hi @agneta1991 & @Ochiengsteven, 

Good job so far!
There are some issues that you still need to work on to go to the next project but you are almost there!



## Highlights

- [ ] Follow Gitflow and GitHub flow ✔️ 
- [ ] Users can reserve rockets and cancel the reservation ✔️ 
- [ ] Users can join the mission and leave the mission ✔️ 
- [ ] Good commit history ✔️ 

## Required Changes ♻️

- [ ] Please kindly check the comments under the review 👇🏾 👇🏾 

## Optional suggestions
_Every comment with the [OPTIONAL] prefix is not crucial enough to stop the approval of this PR. However, I strongly recommend you consider them as they can improve your code._


Cheers, and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please remember to tag me in your question so I can receive the notification.

**_Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise._**

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561) you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

File: src/styles/Rockets.css
Comment (agneta1991) [reply src/styles/Rockets.css:1]: @cynthiainga just a quick question regarding the css part, as in the requirements it does not say that our app has to match the design provided ( as there are no documentation on color codes, paddings, fonts etc. ) It just gives a suggestion that we can use bootstrap. Could you advise us how we can find the exact color codes and all? Thanks

Comment (agneta1991) [reply src/styles/Rockets.css:1]: @cynthiainga could you please respond as soon as you can as we would like to implement the changes, but as we don't know where to get the documentation for the css part, we are being held back. Thank you so much.

Comment (cynthiainga) [reply src/styles/Rockets.css:1]: @agneta1991 I know it's not Figma and they didn’t give the exact color but you should use colors that are close to the given design. Hint `#1278f8 for reserve button`, and `#19a3b9 for reservation`, for padding is just an estimation but you should make sure that there is not much difference.

File: src/redux/rockets/rocketSlice.js [hunk 3]
```diff
@@ -0,0 +1,73 @@
+/* eslint-disable max-len */
```
Comment (Alusp) [inline src/redux/rockets/rocketSlice.js]: - [ ] You have done really well in implementing the changes requested by the previous code reviewer, :clap: , However, as she mentioned, `"Please kindly do not disable linters, fixing linters helps us to write a strong and clean code base"` 😁, Please resolve that.

File: src/redux/rockets/rocketSlice.js [hunk 4]
```diff
@@ -0,0 +1,73 @@
...
+const initialState = {
+  rockets: [],
+  status: 'idle',
+  error: null,
+};
+
+const baseUrl = 'https://api.spacexdata.com/v4/rockets';
+
+export const fetchRocketsAsync = createAsyncThunk('rockets/fetchRockets', async () => {
+  // eslint-disable-next-line no-useless-catch
```
Comment (Alusp) [inline src/redux/rockets/rocketSlice.js]: - [ ] You have done really well in implementing the changes requested by the previous code reviewer, :clap: , However, as she mentioned, `"Please kindly do not disable linters, fixing linters helps us to write a strong and clean code base"` 😁, Please resolve that.

File: src/components/MissionsProfile.css
Comment (Alusp) [inline src/components/MissionsProfile.css:1]: - [ ] You have done really well in implementing the changes requested by the previous code reviewer, :clap: , However, as she mentioned, `"Please kindly do not disable linters, fixing linters helps us to write a strong and clean code base"` 😁, Please resolve that. `"you should make sure your design matches the given design "colors, padding, margin,..."`
matching your design to the given design requirements enhances the chances of creating a successful and well-rounded design solution that meets the needs of users, complies with regulations, and performs effectively and efficiently. :slightly_smiling_face: 

Below is an illustration of yours and the given Design, Please consider the red painting in  your  design :arrow_down: 

<table>
    <tr>
      <td>Your Design</td>
      <td>Given Design</td>
    </tr>
    <tr>
      <td>

student D


</td>
      <td>

space-travelers__missions

</td>
    </tr>
  </table>

**Review by Alusp — CHANGES_REQUESTED** [review]
Hi @agneta1991 and @Ochiengsteven , 

Great work on making the changes requested by a previous reviewer 👏 
You've done well implementing some of the requested changes, but there are still some which aren't addressed yet.

 close

 ## Highlights
 Follow Gitflow and GitHub flow ✔️
 Users can reserve rockets and cancel the reservation ✔️
 Users can join the mission and leave the mission ✔️
 Good commit history ✔️

 ## Required Changes ♻️
_Check the comments under the review._

## Optional suggestions

_Every comment with the [OPTIONAL] prefix won't stop the approval of this PR. However, I strongly recommend you to take them into account as they can make your code better. Some of them were simply missed by the previous reviewer and addressing them will really improve your application._


Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please, remember to tag me @Alusp in your question so I can receive the notification.

**_Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise._**


------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561)  you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

**Review by Ibizugbe — APPROVED** [review]
Hi @Team👋🏼👋🏼👋🏼, 

approved


Your project is complete!  There is nothing else to say other than... it's time to merge it :shipit: 
Congratulations! 🎉 

## Highlights✅✅✅

- [x] Nice work making your project match the given design as requested by the previous reviewers👍🏼.
- [x] Linters are passing👍🏼.

## Optional suggestions

_Every comment with the [OPTIONAL] prefix won't stop the approval of this PR. However, I strongly recommend you to take them into account as they can make your code better. Some of them were simply missed by the previous reviewer and addressing them will really improve your application._

Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please, remember to tag me in your question so I can receive the notification.

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561)  you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

**Pull request merged**